packages/typespec-client-generator-core/src/types.ts

```diff
@@ -1737,6 +1739,37 @@ function updateSpreadModelUsageAndAccess(context: TCGCContext): void {
   }
 }
 
+function handleLegacyHierarchyBuilding(context: TCGCContext): [void, readonly Diagnostic[]] {
+  const diagnostics = createDiagnosticCollector();
+  for (const sdkType of context.__referencedTypeCache.values()) {
+    if (sdkType.kind !== "model" || !sdkType.baseModel) continue;
+    // if the model has legacyHierarchyBuilding, then we should update its discriminated subtypes
+    const legacyHierarchyBuilding = getLegacyHierarchyBuilding(context, sdkType.__raw as Model);
+    // must be done after discriminator is added
+    // Populate discriminated subtypes for legacy hierarchy building
+    if (legacyHierarchyBuilding && sdkType.discriminatorValue) {
+      let currBaseModel: SdkModelType | undefined = sdkType.baseModel;
+      while (currBaseModel) {
+        if (!currBaseModel.discriminatedSubtypes) {
+          currBaseModel.discriminatedSubtypes = {};
+        }
+        currBaseModel.discriminatedSubtypes[sdkType.discriminatorValue] = sdkType;
```

Comment:
In m4, all the layers have `discriminatedProperty`, I'm think of if TCGC needs to copy that?

Reply:
yeah we might as well, let me get that in

Reply:
I'm not sure about how each language deal with multiple layer discriminator. If they rely on discriminated property and treat each layer separately. Then we should add it.

Reply:
Impl in Java should be fine with current behavior (that only the most base class has `discriminatedProperty`) -- impl would find one from its parents, and combine it with `discriminatorValue`